lib/libc/man/fdate.3
.\"
.\" $Id: fdate.3,v 1.3 89/12/26 11:16:38 bww Exp $
.\"
.\" HISTORY
.\" $Log:	fdate.3,v $
.\" Revision 1.3  89/12/26  11:16:38  bww
.\" 	Revised for 2.6 MSD release.
.\" 	[89/12/25            bww]
.\" 
.\" 13-Nov-86  Andi Swimmer (andi) at Carnegie-Mellon University
.\"	Revised for 4.3.
.\"
.\" 11-Sep-84   Leonard Hamey (lgh) at Carnegie-Mellon University
.\"	Documented.
.\"
.TH FDATE 3 9/11/84
.CM 3
.SH "NAME"
fdate \- format date specification
.SH "SYNOPSIS" 
.B
#include <sys/types.h> 
.br
.B
#include <sys/time.h> 
.PP
.br
.B
char *fdate(buffer,format,tm)
.br
.B
char *buffer, *format;
.br
.B
struct tm *tm;
.SH "DESCRIPTION"
.I
Fdate
formats the information supplied in a
.I
tm
structure
into an ASCII date specification. The formatting is controlled by the
.I
format
string (described below).
The result is stored in
.IR buffer ,
and a pointer to it returned as the function value.
.I
Buffer
should be pre-allocated to be large enough for the result string.
.SH "FORMATS"
This section describes the format specification options which
may be used in the
.I format
string.
These patterns resemble those used
by
.IR printf (3).
.PP
The percent sign indicates that a field of the
.I
tm
structure
is to be inserted into the result at that point.
The fields are named as follows:
.PP
.RS
.PD 0
.TP 10
%wday
name of the week day
.TP 10
%day
number of the month day
.TP 10
%month
name of the month
.TP 10
%nmonth
number of the month
.TP 10
%year
year
.TP 10
%yday
number of the day within the year
.TP 10
%hour
hour of the day
.TP 10
%min
minutes
.TP 10
%sec
seconds
.TP 10
%noon
generates the word "noon" at noon, fails otherwise
.TP 10
%midnight
generates "00:00" at midnight, fails otherwise
.TP 10
%am
generates "am" or "pm" as appropriate
.TP 10
%time
generates the time in a format explained below
.TP 10
%th
.B
st, nd, rd
or
.B
th
as appropriate.
.RE
.PD
.PP
Simply naming the field will insert it in the result. Thus,
.IP
.B
"Today is %wday"
.PP
will result in
.IP
.B
"Today is tuesday"
.PP
on Tuesday. Of course, it would be better if the weekday was indicated with a
capital letter. The alphabetic fields are
capitalised in the same manner as the field was named, so
.B
%Wday
inserts
.B
Tuesday
and
.B
%WDAY
inserts
.B
TUESDAY.
.PP
You can also specify variable formats, allowing the format of the result to
vary depending on the date/time being formatted.  This is done by surrounding
part of the format string in brackets or braces.  If a format string is
enclosed
in square brackets, then the entire string will produce no result if any
field enclosed within the brackets itself failed to produce a result.  A
field will fail to produce a result if it is unspecified in the date and cannot
be computed from the date (e.g. when the time is not specified).
.PP
Thus, for example, in the format string,
.IP
.B
"hello {how are you %sec %min} goodbye"
.PP
the result will be simply
.IP
.B
"hello goodbye"
.PP
if either seconds or minutes
are undefined.  Otherwise, it will result in
something like
.IP
.B
"hello how are you 34 23 goodbye".
.PP
Square
brackets function almost identically, except that the
entire string is not inserted only if
.B
all
of the fields are undefined.  Thus, 
.IP
.B
"hello [how are you %sec %min] goodbye"
.PP
will result in
.IP
.B
"hello goodbye"
.PP
if both seconds
and minutes are undefined, but will otherwise insert the middle part as well.
.PP
Within brackets, alternatives can be specified by separating them with vertical
bars ("|").  The first alternative to succeed is the one that is used.
.PP
While %field generally gives the value of the field, %?field may
be used to query the
nature of the field; if it has a reasonable default
value, then the field fails.  Reasonable defaults are zero for the minutes and
seconds, and the current year.
.PP
Additionally, a numeric field width may be specified between the % and the
field name.  If the field width begins with a zero, the field will be
inserted with leading zeroes. For the hour field, a leading zero also serves to
indicate 24-hour notation.
.SH "EXAMPLES"
As a simple case, the 
.I format
string
.IP
.B
"Today is the %yday%th day of %year."
.PP
will result in something like
.IP
.B
"Today is the 101st day of 1987."
.PP
Similarly,
.IP
.B
"%3Wday %2day %3Month %year %0hour:%min:%sec %yday"
.PP
will result in something like
.IP
.B
"Sat  7 Apr 1984 16:32:00 97"
.PP
As a final, more complex example, if the format given is:
.IP
.B
"[%midnight|%noon|%0hour:%min:%?sec|{%hour{:%?min}%am}]"
.PP
This will insert "00:00" if it is midnight, or "noon" if
it is noon.  Otherwise,
if the seconds are defined it will insert the hour (in 24-hour format
because of the field width zero), minute, and second, separated by colons.
If the seconds are not defined or are equal to zero, it will insert the hour (in
12 hour format), the minutes if they are defined (omitting the relevant colon
if they aren't defined), and "am" or "pm".  This format, incidentally, is
precisely what you can get with the rather simpler field "%time".
.SH "SEE ALSO"
.IR whenis (1),
.IR time (2),
.IR ctime (3),
.IR gtime (3),
.IR localtime (3),
.IR parsedate (3)
.SH "DIAGNOSTICS"
There are no error conditions. An invalid field name will be treated as
literal text.


X11/CHANGES
The following changes have been made to this directory since R3:

     o	the configuration files have moved from util/imake.includes/ to config/
     o	sources for imake are now in config/ instead of util/imake/.



lib/libc/compat-cmu/fdate.c
# include <sys/types.h>
# include <sys/time.h>

/* 
 *  HISTORY
 * 30-Apr-85  Steven Shafer (sas) at Carnegie-Mellon University
 *	Adapted for 4.2 BSD UNIX:  macro file name changed.
 *
 * 
 *  March 1984 - Leonard Hamey (lgh) at Carnegie-Mellon University
 *     Created.
 */

# define TEXT 1
# define FAIL 2
# define OPTIONAL 3
# define SUCCEED 4

# define YES 1
# define NO 0
# define MAYBE 2

# define SMONTHS 0
# define SWDAYS 12
# define SNOON 19
# define SMIDNIGHT 20
# define SAM 21
# define SPM 22
# define STH 23
# define SST 24
# define SND 25
# define SRD 26

# define NOON 0
# define HOUR 1
# define AM 2
# define MIN 3
# define SEC 4
# define WDAY 5
# define DAY 6
# define TH 7
# define NMONTH 8
# define MONTH 9
# define YEAR 10
# define YDAY 11
# define MIDNIGHT 12
# define TIME 13

static char *strings[] =
{ "jan*uary", "feb*ruary", "mar*ch", "apr*il", "may", "jun*e",
  "jul*y", "aug*ust", "sep*tember", "oct*ober", "nov*ember", "dec*ember",
  "sun*day", "mon*day", "tue*s*day", "wed*nesday", "thu*r*s*day",
  "fri*day", "sat*urday",
  "noon", "00:00", "am", "pm", "th", "st", "nd", "rd"
};

static char *keys[] =
{ "noon", "hour", "am", "min", "sec",
  "wday", "day", "th", "nmonth", "month", "year", "yday",
  "midnight", "time",
  0
};

/* The following is a macro definition of %time. */

static char *mactime =
  "[%midnight|%noon|%0hour:%min:%?sec|{%hour{:%?min}%am}]";

static char *p;
static int resp;
static char *result;

static dofield ();
static percent ();
static char foldc();
static formfld ();
static copyword ();
static casefix ();

char *fdate (resb, patt, tm)
char *resb, *patt;
struct tm *tm;
{ result = resb;
  resp = 0;
  p = patt;
  dofield (' ', tm);
  result[resp] = '\0';
  return (result);
}

static dofield (doing, tm)
char doing;
struct tm *tm;
{ int ndep, s, sresp, sresp2;
  int fail;
  char ch;
  fail = MAYBE;
  sresp = resp;
  while (*p)
  { if (*p == '{' || *p == '[')
    { ch = *p++;
      s = dofield (ch, tm);
      if (s == FAIL && doing == '[')
	fail = YES;
      else if (s == SUCCEED && doing == '{')
	fail = NO;
    }
    else if (*p == '}' || *p == ']' || *p == '|')
    { if (doing == '{' && fail == MAYBE)
	fail = YES;
      if (fail == YES)
      { resp = sresp;
        if (*p == '|')
        { p++;
	  fail = MAYBE; 			     /* try next */
	  continue;
        }
	p++;
      }
      else if (*p == '|')
      { /* Scan to matching bracket/brace */
	ndep = 1;
	while (ndep > 0)
	{ if (*p == '{' || *p == '[')
	    ndep++;
	  else if (*p == '}' || *p == ']')
	    ndep--;
	  p++;
	}
      }
      else
	p++;
      return (fail == YES ? FAIL : SUCCEED);
    }
    else if (*p == '%')
    { sresp2 = resp;
      s = percent (tm);
      if (doing == '{')
      { if (s == FAIL)		     /* Discard failure results */
	  resp = sresp2;
	else if (s == SUCCEED)
	  fail = NO;
      }
      else if (doing == '[')
      { if (s == FAIL)
	  fail = YES;
      }
    }
    else
    { result[resp++] = *p;
      p++;
    }
    if (fail == YES)
    { /* Failure: scan for barline or closing bracket/brace; leave ptr there */
      resp = sresp;
      ndep = 1;
      while (ndep > 0)
      { if (*p == '|' && ndep == 1)
	  break;
	if (*p == '[' | *p == '{')
	  ndep++;
	else if (*p == ']' || *p == '}')
	{ ndep--;
	  if (ndep == 0)
	    return (FAIL);
	}
	p++;
      }
    }
  }
  return (SUCCEED);
}

static percent (tm)
struct tm *tm;
{ register char *pp;
  char *spp;
  int query, lead0, prec, upcase, sresp, s;
  char *savep;
  register int ks;
  register char *k;
  char foldc ();
  int formfld ();
  char *fldp;
  pp = p + 1;
  if (*pp == '%' || *pp == '{' || *pp == '}' || *pp == '|' ||
    *pp == '[' || *pp == ']')
  { p += 2;
    result[resp++] = *pp;
    return (TEXT); 			     /* Was just text */
  }
  query = *pp == '?'; 			     /* Query? */
  if (query)
    pp++;
  lead0 = *pp == '0';	 		     /* Leading zero? */
  if (lead0)
    pp++;
  if (*pp >= '1' && *pp <= '9') 	     /* Precision? */
  { prec = *pp - '0';
    pp++;
  }
  else
    prec = 0;
  upcase = 0;	 			     /* Case? */
  if (*pp >= 'A' && *pp <= 'Z')
    upcase = 1;
  if (pp[1] >= 'A' && pp[1] <= 'Z')
    upcase = 2;

  spp = pp;
  for (ks = 0; keys[ks]; ks++) 	     /* Check for keyword */
  { for (k = keys[ks], pp = spp; *k; k++, pp++)
      if (foldc (*k) != foldc (*pp))
        break;
    if (! *k)	 			     /* Match found */
      break;
  }
  if (! keys[ks]) 			     /* No match */
  { result[resp++] = '%'; 		     /* Treat as text */
    p++;
    return (TEXT);
  }
  p = pp;
  if (ks == TIME) 			     /* Macro */
  { savep = p;
    sresp = resp;
    p = mactime + 1; 			     /* Skip leading bracket */
    s = dofield (mactime[0], tm);
    p = savep;
    return (s);
  }
  /* Match found */
  s = formfld (tm, ks, lead0, prec, &fldp);
  sresp = resp;
  if (s == OPTIONAL)
  { if (query)
      s = FAIL;
    else
      s = SUCCEED;
  }
  if (s == SUCCEED)
  { for (; *fldp; fldp++)
      result[resp++] = *fldp;
    casefix (&result[sresp], upcase);
  }
  return (s);
}

static char foldc (c)
char c;
{ return (c < 'a' || c > 'z' ? c : c - 'a' + 'A');
}

static char nstr[30];
static char sfield[30];
static int thmem;

static formfld (tm, field, lead0, prec, result)
struct tm *tm;
int field, lead0, prec;
char **result;
{ int fld, ddiff, s;
  struct tm ctm, *localtime ();
  long ctime, time ();
  sfield[0] = '\0';
  *result = sfield;
  if (field != TH)
    thmem = -1;
  switch (field)
  { case NOON:
      if (tm->tm_hour == 12 && tm->tm_min == 0 && tm->tm_sec == 0)
      { *result = strings[SNOON];
	return (SUCCEED);
      }
      return (FAIL);

    case HOUR:
      for (; prec > 2; prec--)
	strcat (sfield, " ");
      if (lead0)
        sprintf (nstr, "%02d", tm->tm_hour);
      else if (prec < 2)
	sprintf (nstr, "%d", (tm->tm_hour + 11) % 12 + 1);
      else
	sprintf (nstr, "%2d", (tm->tm_hour + 11) % 12 + 1);
      strcat (sfield, nstr);
      thmem = tm->tm_hour;
      return (tm->tm_hour >= 0 ? SUCCEED : FAIL);

    case MIN:
      fld = tm->tm_min;
    case SEC:
      if (field == SEC)
	fld = tm->tm_sec;
      if (prec == 0)
	prec = 2;
      for (; prec > 2; prec--)
	strcat (sfield, " ");
      if (prec < 2)
	sprintf (nstr, "%d", fld);
      else
	sprintf (nstr, "%02d", fld);
      strcat (sfield, nstr);
      thmem = fld;
      return (fld > 0 ? SUCCEED : (fld == 0 ? OPTIONAL : FAIL));

    case AM:
      if (tm->tm_hour >= 12)
	*result = strings[SPM];
      else if (tm->tm_hour >= 0)
	*result = strings[SAM];
      return (tm->tm_hour >= 0 ? SUCCEED : FAIL);

    case WDAY:
      if (tm->tm_wday >= 0 && tm->tm_wday < 7)
      { copyword (sfield, strings[SWDAYS + tm->tm_wday], prec);
	return (SUCCEED);
      }
      return (FAIL);

    case MONTH:
      if (tm->tm_mon >= 0 && tm->tm_mon < 12)
      { copyword (sfield, strings[SMONTHS + tm->tm_mon], prec);
	return (SUCCEED);
      }
      return (FAIL);

    case DAY:
      fld = tm->tm_mday;
    case NMONTH:
    case YDAY:
      if (field == NMONTH)
	fld = tm->tm_mon + 1;
      else if (field == YDAY)
	fld = tm->tm_yday;
      if (prec == 0)
	prec = 1;
      if (lead0)
      { strcpy (nstr, "%09d");
	nstr[2] = prec + '0';
      }
      else
      { strcpy (nstr, "%9d");
	nstr[1] = prec + '0';
      }
      sprintf (sfield, nstr, fld);
      thmem = fld;
      return (fld >= 0 ? SUCCEED : FAIL);

    case TH:
      if (thmem >= 0)
      { if ((thmem / 10) % 10 == 1)
	  *result = strings[STH];
	else if (thmem % 10 == 1)
	  *result = strings[SST];
	else if (thmem % 10 == 2)
	  *result = strings[SND];
	else if (thmem % 10 == 3)
	  *result = strings[SRD];
	else
	  *result = strings[STH];
	return (SUCCEED);
      }
      return (FAIL);

    case YEAR:
      ctime = time (0);
      ctm = *localtime (&ctime);
      ddiff = (tm->tm_year - ctm.tm_year) * 366 + tm->tm_yday - ctm.tm_yday;
      if (tm->tm_year >= -1900)
      { if (prec == 0 || prec > 3)
	  sprintf (nstr, "%04d", tm->tm_year + 1900);
	else
	  sprintf (nstr, "%02d", (tm->tm_year + 1900) % 100);
	*result = nstr;
	return (ddiff < 0 || ddiff > 300 ? SUCCEED : OPTIONAL);
      }
      return (FAIL);

    case MIDNIGHT:
      if (tm->tm_hour == 0 && tm->tm_min == 0 && tm->tm_sec == 0)
      { *result = strings[SMIDNIGHT];
	return (SUCCEED);
      }
      return (FAIL);
  }
}

/* copyword: Copy <from> to <to> subject to precision <prec>. Asterisks in
 * <from> mark valid truncation points. The longest string no longer than
 * <prec> and broken at a valid truncation point is returned in <to>. Note
 * that the result may exceed <prec> in length only if there are no valid
 * truncation points short enough. The shortest is then taken.
 * 
 * e.g. Tue*s*day.  Precision 1-3 -> Tue,  Precision 4-6 -> Tues,
 *      Precision 0 and 7... -> Tuesday.
 */

static copyword (to, from, prec)
char *to, *from;
int prec;
{ int ast = 0, top = 0;
  if (prec == 0)
    prec = 999;
  for (; *from; from++)
  { if (*from == '*')
      ast = top;
    else
    { to[top] = *from;
      top++;
    }
    if (top > prec && ast > 0)
    { /* Required precision exceeded and asterisk found */
      to[ast] = '\0'; 			     /* Truncate at asterisk */
      return;
    }
  }
  to[top] = '\0';
  return;
}

/* casefix: select case of word. 1: first letter raised. 2: all raised. */

static casefix (text, upcase)
char *text;
int upcase;
{ if (upcase == 1)
    *text = foldc (*text);
  else if (upcase > 0)
    foldup (text, text);
}
